File: Copilot/Java/Q7.java
package Java;

public class Q7 {
    public int countPrimes(int n) {
        //Given an integer n, return the number of prime numbers 
        //that are strictly less than n. with 0 <= n <= 5 * 106
    
    
    /* In questo primo caso, il programma non funziona perchè non è stato implementato il metodo isPrime
        int count = 0;
        for (int i = 2; i < n; i++) {
            if (isPrime(i)) {
                count++;
            }
        }
        return count;
    }
    private boolean isPrime(int i) {
        return false;
    }*/



    /*2° tentativo, il codice eseguiva correttamente i test ma andava contro a time limit exceeded
        int count = 0;
        for (int i = 2; i < n; i++) {
            if (isPrime(i)) {
                count++;
            }
        }
        return count;
    }
    public boolean isPrime(int n) {
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0) {
                return false;
            }
        }
        return true;
     */
    

      //Esecuzione corretta, 3° tentativo
        boolean[] isPrime = new boolean[n];
        for (int i = 2; i < n; i++) {
            isPrime[i] = true;
        }
        for (int i = 2; i * i < n; i++) {
            if (!isPrime[i]) {
                continue;
            }
            for (int j = i * i; j < n; j += i) {
                isPrime[j] = false;
            }
        }
        int count = 0;
        for (int i = 2; i < n; i++) {
            if (isPrime[i]) {
                count++;
            }
        }
        return count;
        
    }
    

    
}

File: Copilot/JavaScript/Q7.js
/**
 * @param {number} n
 * @return {number}
 */
var countPrimes = function(n) {
    //Given an integer n, return the number of prime numbers that are strictly less than n.

    //Example 1:
    //Input: n = 10
    //Output: 4
    //Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.

    //Example 2:
    //Input: n = 0
    //Output: 0

    //Example 3:
    //Input: n = 1
    //Output: 0

    //Constraints:
    //0 <= n <= 5 * 106

    //SOLUTION 13
    let count = 0;
    let isPrime = new Array(n).fill(true);
    for (let i = 2; i < n; i++) {
        if (isPrime[i]) {
            count++;
            for (let j = i * i; j < n; j += i) {
                isPrime[j] = false;
            }
        }
    }
    return count;


};